### DISCUSSION TITLE
Wrong Response when browsing through structure

### DISCUSSION THREAD

**Discussion by BenediktEnger** [FreeOpcUa/python-opcua#1461] [Q&A]
Hi,
I try to get an FreeOpcUa Server to run in combination with a B&R PLC.
I have similar Server developed with .NET and this one is running fine.

So I ran into an issue:
When the client wants to get the node Handle from the server it fails with error 8034_0000.
I tracked the communication and compared it with the .NET server. 
The difference between these two is, that after requesting the encoding for the type inside the output argument, the FreeOpcUa Server returns an empty array. 
![image](https://user-images.githubusercontent.com/44339476/166668694-8b47d3e2-62f1-4a64-aa4a-b9cfd09f294a.png)

The .NET server is returning the node of the description of the encoding type.

![image](https://user-images.githubusercontent.com/44339476/166668908-b83d664b-efcd-4506-8d80-29b338d95443.png)

Does somebody know how to fix that?

If you need the full wireshark  record, let me know.

EDIT: I had a deeper look on the wireshark record and the client is requesting a description of the encoding which seems not to be present. When following the references in UAExpert I get this node:

![image](https://user-images.githubusercontent.com/44339476/166703832-2b7f2111-6ecf-491d-8286-1a1b3f79db50.png)

Instead of the node at the OPCBinaries Object:
![image](https://user-images.githubusercontent.com/44339476/166703974-172f6823-12aa-46c4-96e6-9783ccc3dfc3.png)

**Comment by schroeder-** [comment]
BadNodeIdUnknown means the server doesn't have the node id you requested. Can you check via UAExpert if a Node with the NodeId exists?

**Reply by BenediktEnger** [reply]
The nodeId exists. When checking the client requests I can see that the method id is requested and responded the right way. Otherwise it would not get to the point to request the encoding of the datatype of the outputArgument

**Reply by BenediktEnger** [reply]
I double checked it and the response to the method node is good:
![image](https://user-images.githubusercontent.com/44339476/166697184-bd055631-5fd0-4d69-9e69-3adad5bed173.png)
![image](https://user-images.githubusercontent.com/44339476/166697260-4c15326e-4016-4ad9-8f7a-8a0d3079b76f.png)

**Reply by schroeder-** [reply]
Wireshark doesn't help here, problem is not the communication. Its more a problem in the addresspace. Can you provide a minimal example how you create the the variable in question?

**Reply by BenediktEnger** [reply]
Sure.
I do this by an recursiv function. The output argument is a nested struct, so  the function is called twice.

```
def addStruct(self, server):
        nodeId = ua.extension_object_ids.get(self.typeName)
        if nodeId != None:
            return nodeId
                
        struct = server.dict_builder.create_data_type(self.typeName)
         
        for member in self.members:
            if issubclass(type(member.type), DataTypeType):
                member.type.addStruct(server)
                struct.add_field(member.name, member.type.typeName)
            else:
                struct.add_field(member.name, member.type)

        server.dict_builder.set_dict_byte_string()
        server.server.load_type_definitions()
        return struct.data_type
```

Have you seen my edit on the question?

**Reply by schroeder-** [reply]
Ok you create custom structs in OPC UA 1.03 format and your clients wants customs structs in V1.04 format. I don't know if pythonua supports this, in asyncua this is supported. See this example https://github.com/FreeOpcUa/opcua-asyncio/blob/259811d5ad11c47c8d2b2289a8a2783b97c79421/examples/server-custom-structures-amd-enums.py. Also if you don't want to switch to async, there is a sync wrapper for this case.

**Reply by BenediktEnger** [reply]
I switched to asyncio and the error is not gone. When the client asks for encoding of the type, the server returns Array with length 0 (so I guess not available). Strange thing is, watching the type with uaExpert, the type node has an encoding node. 

I saw that there is no Dictionary. Using pythonua I created types using the dictionarybuilder. This seems not necessary anymore but then there is no dictionary on the server which gives the encoding the description

**Reply by schroeder-** [reply]
did you switch your code to:
```
from asyncua.common.structures104 import new_struct, new_enum, new_struct_field
custom_struct = new_struct(server, idx, ....)
```
instead of dict_builder?

**Reply by BenediktEnger** [reply]
Yes. use the sync wrapper

```
from asyncua.sync import new_enum, new_struct, new_struct_field

structFields = []
for member in self.members:
    if issubclass(type(member.type), DataTypeType):
        nodeid = member.type.addStruct(server)
        structFields.append(new_struct_field(member.name, nodeid))
    else:
        structFields.append(new_struct_field(member.name, member.type))

struct = new_struct(server.server, server.idx, self.typeName,structFields)
```

**Reply by schroeder-** [reply]
Sry I think I know understand the problem, first the node the client want to read a component of a dictionary, which is pre 1.04 structs, sry for the misunderstanding. So you want your first code. Maybe there is a fix in asyncua, if this doesn't work someone has to fix the dictionary creation. I never touched pr 1.04 stuff and I never needed it, hope someone can help you here.

**Reply by BenediktEnger** [reply]
I‘m not sure of the client ist demanding v1.04. As I added to the question, my guess is that the link between the encoding and description is broken. This is what I can see in UaExpert. 
I already double checked the adding of the references in the code (in V1.03) but I can not find a bug here

**Comment by BenediktEnger** [comment]
I just digged a bit deeper in what is happening:

1. As I added in the question, the reference to the description is broken. This is because the reference_generator always adds TargetNodeClass = ua.NodeClass.DataType which is wrong in this case. The Description is of type Variable.
So I changed the method _link_nodes and _reference_generator to the following:


```
 def _link_nodes(self, linked_obj_node_id, data_type_node_id, description_node_id):
        """link the three node by their node ids according to UA standard"""
        refs = [
                # add reverse reference to BaseDataType -> Structure
                _reference_generator(data_type_node_id, ua.NodeId(ua.ObjectIds.Structure, 0),
                                     ua.NodeId(ua.ObjectIds.HasSubtype, 0), ua.NodeClass.DataType, False),
                # add link reference to encription
                _reference_generator(data_type_node_id, linked_obj_node_id,
                                     ua.NodeId(ua.ObjectIds.HasEncoding, 0), ua.NodeClass.Object),
                # add reverse reference to created data type
                _reference_generator(linked_obj_node_id, data_type_node_id,
                                     ua.NodeId(ua.ObjectIds.HasEncoding, 0), ua.NodeClass.DataType, False),
                # add HasDescription link to dictionary description
                _reference_generator(linked_obj_node_id, description_node_id,
                                     ua.NodeId(ua.ObjectIds.HasDescription, 0), ua.NodeClass.Variable),
                # add reverse HasDescription link
                _reference_generator(description_node_id, linked_obj_node_id,
                                     ua.NodeId(ua.ObjectIds.HasDescription, 0), ua.NodeClass.Object, False),
                # add link to the type definition node
                _reference_generator(linked_obj_node_id, ua.NodeId(ua.ObjectIds.DataTypeEncodingType, 0),
                                     ua.NodeId(ua.ObjectIds.HasTypeDefinition, 0), ua.NodeClass.Object),
                # add has type definition link
                _reference_generator(description_node_id, ua.NodeId(ua.ObjectIds.DataTypeDescriptionType, 0),
                                     ua.NodeId(ua.ObjectIds.HasTypeDefinition, 0), ua.NodeClass.Variable),
                # add forward link of dict to description item
                _reference_generator(self.dict_id, description_node_id,
                                     ua.NodeId(ua.ObjectIds.HasComponent, 0),ua.NodeClass.Variable),
                # add reverse link to dictionary
                _reference_generator(description_node_id, self.dict_id,
                                     ua.NodeId(ua.ObjectIds.HasComponent, 0), ua.NodeClass.DataType, False)]
        self._session_server.add_references(refs)

def _reference_generator(source_id, target_id, reference_type, target_node_class, is_forward=True):
    ref = ua.AddReferencesItem()
    ref.IsForward = is_forward
    ref.ReferenceTypeId = reference_type
    ref.SourceNodeId = source_id
    ref.TargetNodeClass = target_node_class
    ref.TargetNodeId = target_id
    return ref
```

Now the linking works.

2. Now, watching the wireshark trace, I saw, that the client is requesting first the datatype, then the encoding and then the description. When requesting the description, the client asks for TypeDefinition:
![image](https://user-images.githubusercontent.com/44339476/167105007-034b5e04-f967-46db-a381-403a37f36e79.png)

but the response from the python server is 0 even if he has this TypeDefinition as I saw using UAExpert.
![image](https://user-images.githubusercontent.com/44339476/167105197-1676c9d2-be30-4601-a22b-78b10d8cb5b3.png)

![image](https://user-images.githubusercontent.com/44339476/167105583-60b07157-37b2-4821-802c-09b6b2fdad9f.png)



So, does anybody had this issue before or knows how to solve it?
Thanks!
